File: backend/internal/util/key_generator.go
package util

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"strconv"
	"sync"
	"time"
)

const (
	TargetKeyLength = 32
	TimestampLength = 8
	CounterLength   = 4
	MaxPrefixLength = 20
	base36Chars     = "0123456789abcdefghijklmnopqrstuvwxyz"
)

// KeyGenerator generates unique 32-character redemption code keys
// Structure: [prefix][random_part][timestamp_base36][counter_base36]
type KeyGenerator struct {
	counter int64
	mu      sync.Mutex
}

var defaultKeyGen = &KeyGenerator{}

// GenerateKey creates a single unique 32-character key
func GenerateKey(prefix string) (string, error) {
	return defaultKeyGen.Generate(prefix)
}

// GenerateBatch creates a batch of unique keys
func GenerateBatch(count int, prefix string) ([]string, error) {
	return defaultKeyGen.GenerateBatch(count, prefix)
}

// Generate creates a single unique 32-character key
func (g *KeyGenerator) Generate(prefix string) (string, error) {
	if len(prefix) > MaxPrefixLength {
		return "", fmt.Errorf("prefix must not exceed %d characters", MaxPrefixLength)
	}

	// Calculate random part length
	randomLength := TargetKeyLength - len(prefix) - TimestampLength - CounterLength
	if randomLength < 8 {
		randomLength = 8
	}

	// Generate random part
	randomBytes := make([]byte, (randomLength/2)+1)
	if _, err := rand.Read(randomBytes); err != nil {
		return "", fmt.Errorf("failed to generate random bytes: %w", err)
	}
	randomPart := hex.EncodeToString(randomBytes)[:randomLength]

	// Timestamp in base36 (milliseconds)
	timestampMs := time.Now().UnixMilli()
	timestampB36 := Base36Encode(timestampMs)
	if len(timestampB36) > TimestampLength {
		timestampB36 = timestampB36[len(timestampB36)-TimestampLength:]
	}
	for len(timestampB36) < TimestampLength {
		timestampB36 = "0" + timestampB36
	}

	// Counter in base36
	g.mu.Lock()
	g.counter = (g.counter + 1) % 1679616 // 36^4
	counterVal := g.counter
	g.mu.Unlock()

	counterB36 := Base36Encode(counterVal)
	if len(counterB36) > CounterLength {
		counterB36 = counterB36[len(counterB36)-CounterLength:]
	}
	for len(counterB36) < CounterLength {
		counterB36 = "0" + counterB36
	}

	// Combine
	key := prefix + randomPart + timestampB36 + counterB36

	// Ensure exactly 32 characters
	if len(key) > TargetKeyLength {
		key = key[:TargetKeyLength]
	} else if len(key) < TargetKeyLength {
		for len(key) < TargetKeyLength {
			key += "0"
		}
	}

	return key, nil
}

// GenerateBatch creates a batch of unique keys
func (g *KeyGenerator) GenerateBatch(count int, prefix string) ([]string, error) {
	if count < 1 || count > 1000 {
		return nil, fmt.Errorf("count must be between 1 and 1000")
	}
	if len(prefix) > MaxPrefixLength {
		return nil, fmt.Errorf("prefix must not exceed %d characters", MaxPrefixLength)
	}

	keySet := make(map[string]struct{})
	keys := make([]string, 0, count)
	maxAttempts := count * 3

	for len(keys) < count && maxAttempts > 0 {
		key, err := g.Generate(prefix)
		if err != nil {
			return nil, err
		}
		if _, exists := keySet[key]; !exists {
			keySet[key] = struct{}{}
			keys = append(keys, key)
		}
		maxAttempts--
	}

	if len(keys) < count {
		return nil, fmt.Errorf("failed to generate %d unique keys", count)
	}

	return keys, nil
}

// Base36Encode encodes an integer to base36 string
func Base36Encode(n int64) string {
	if n == 0 {
		return "0"
	}
	if n < 0 {
		return "-" + Base36Encode(-n)
	}
	return strconv.FormatInt(n, 36)
}

File: backend-py/app/key_generator.py
"""
Key Generator module for NewAPI Middleware Tool.
Generates unique 32-character redemption code keys.
"""
import secrets
import time
from typing import List


def base36_encode(number: int) -> str:
    """
    Encode an integer to base36 string.
    
    Args:
        number: Non-negative integer to encode.
        
    Returns:
        Base36 encoded string (lowercase).
    """
    if number < 0:
        raise ValueError("Number must be non-negative")
    if number == 0:
        return "0"
    
    chars = "0123456789abcdefghijklmnopqrstuvwxyz"
    result = []
    while number > 0:
        result.append(chars[number % 36])
        number //= 36
    return "".join(reversed(result))


class KeyGenerator:
    """
    Generator for unique 32-character redemption code keys.
    
    Key structure: [prefix][random_part][timestamp_base36][counter_base36]
    - prefix: User-specified prefix (0-20 chars)
    - random_part: Random bytes converted to hex (variable length)
    - timestamp_base36: Timestamp in base36 (8 chars)
    - counter_base36: Performance counter in base36 (4 chars)
    """
    
    TARGET_LENGTH = 32
    TIMESTAMP_LENGTH = 8
    COUNTER_LENGTH = 4
    MAX_PREFIX_LENGTH = 20
    
    def __init__(self):
        """Initialize the KeyGenerator."""
        self._counter = 0
    
    def generate_key(self, prefix: str = "") -> str:
        """
        Generate a unique 32-character key.
        
        Args:
            prefix: Optional prefix for the key (max 20 chars).
            
        Returns:
            A 32-character unique key string.
            
        Raises:
            ValueError: If prefix exceeds 20 characters.
        """
        if len(prefix) > self.MAX_PREFIX_LENGTH:
            raise ValueError(f"Prefix must not exceed {self.MAX_PREFIX_LENGTH} characters")
        
        # Calculate random part length
        random_length = max(
            8, 
            self.TARGET_LENGTH - len(prefix) - self.TIMESTAMP_LENGTH - self.COUNTER_LENGTH
        )
        
        # Generate random part using secure random bytes
        random_part = secrets.token_hex(random_length // 2 + 1)[:random_length]
        
        # Generate timestamp part (milliseconds since epoch in base36)
        timestamp_ms = int(time.time() * 1000)
        timestamp_b36 = base36_encode(timestamp_ms)[-self.TIMESTAMP_LENGTH:].zfill(self.TIMESTAMP_LENGTH)
        
        # Generate counter part (performance counter modulo to fit in 4 chars)
        # Using combination of perf_counter and internal counter for uniqueness
        perf_value = int(time.perf_counter() * 1000000) % 1679616  # 36^4 = 1679616
        self._counter = (self._counter + 1) % 1679616
        counter_value = (perf_value + self._counter) % 1679616
        counter_b36 = base36_encode(counter_value).zfill(self.COUNTER_LENGTH)[-self.COUNTER_LENGTH:]
        
        # Combine all parts
        key = f"{prefix}{random_part}{timestamp_b36}{counter_b36}"
        
        # Ensure exactly 32 characters
        if len(key) > self.TARGET_LENGTH:
            key = key[:self.TARGET_LENGTH]
        elif len(key) < self.TARGET_LENGTH:
            key = key.ljust(self.TARGET_LENGTH, "0")
        
        return key
    
    def generate_batch(self, count: int, prefix: str = "") -> List[str]:
        """
        Generate a batch of unique keys.
        
        Args:
            count: Number of keys to generate (1-1000).
            prefix: Optional prefix for all keys (max 20 chars).
            
        Returns:
            List of unique 32-character keys.
            
        Raises:
            ValueError: If count is not in range 1-1000 or prefix exceeds 20 chars.
        """
        if not 1 <= count <= 1000:
            raise ValueError("Count must be between 1 and 1000")
        if len(prefix) > self.MAX_PREFIX_LENGTH:
            raise ValueError(f"Prefix must not exceed {self.MAX_PREFIX_LENGTH} characters")
        
        keys = set()
        attempts = 0
        max_attempts = count * 3  # Allow some retries for collision handling
        
        while len(keys) < count and attempts < max_attempts:
            key = self.generate_key(prefix)
            keys.add(key)
            attempts += 1
        
        if len(keys) < count:
            raise RuntimeError(f"Failed to generate {count} unique keys after {max_attempts} attempts")
        
        return list(keys)


# Global instance for convenience
_key_generator: KeyGenerator | None = None


def get_key_generator() -> KeyGenerator:
    """Get or create the global KeyGenerator instance."""
    global _key_generator
    if _key_generator is None:
        _key_generator = KeyGenerator()
    return _key_generator
